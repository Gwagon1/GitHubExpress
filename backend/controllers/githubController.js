const axios = require('axios');

// Fetch user details 
const getUserDetails = async (req, res) => {
  const { username } = req.params;

  if (!username) {
    return res.status(400).json({ error: 'Username is required' });
  }

  try {
    const response = await axios.get(`https://api.github.com/users/${username}`);
    res.json(response.data);
  } catch (error) {
    console.error('Error fetching user details:', error.response?.data || error.message);
    res.status(500).json({ error: 'Error fetching user details' });
  }
};

// Fetch user repositories
const getUserRepos = async (req, res) => {
  const { username } = req.params;

  try {
    const response = await axios.get(`https://api.github.com/users/${username}/repos`);
    res.json(response.data);
  } catch (error) {
    console.error('Error fetching user repositories:', error.response?.data || error.message);
    res.status(500).json({ error: 'Error fetching user repositories' });
  }
};

// Fetch commits for a specific repository
const getRepoCommits = async (req, res) => {
  const { username, repoName } = req.params;

  try {
    const response = await axios.get(`https://api.github.com/repos/${username}/${repoName}/commits`);
    res.json(response.data);
  } catch (error) {
    console.error('Error fetching repository commits:', error.response?.data || error.message);
    res.status(500).json({ error: 'Error fetching repository commits' });
  }
};

module.exports = {
  getUserDetails,
  getUserRepos,
  getRepoCommits,
};
