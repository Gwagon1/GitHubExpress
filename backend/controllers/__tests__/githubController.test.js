import axios from 'axios';
import { getUserDetails } from '../githubController';


// Mock axios.get method
jest.mock('axios');


describe('getUserDetails', () => {
  it('returns a list of GitHub users', async () => {
    const mockRes = { json: jest.fn() };
    const mockReq = { params: { username: 'octocat' } };

    
    // Mock the response of axios.get to return a fake response for the users search
    axios.get.mockResolvedValue({ data: { login: 'octocat', name: 'Octocat' } });


    // Call the controller function
    await getUserDetails(mockReq, mockRes);


    // Verify axios.get was called with the correct query
    expect(axios.get).toHaveBeenCalledWith('https://api.github.com/users/octocat');
    expect(mockRes.json).toHaveBeenCalledWith({ login: 'octocat', name: 'Octocat' });
  });


  it('handles errors gracefully', async () => {
    const mockRes = { status: jest.fn().mockReturnThis(), json: jest.fn() };
    const mockReq = { params: { username: 'invalid' } };


    // Mock axios.get to throw an error
    axios.get.mockRejectedValue(new Error('GitHub API error'));


    // Call the controller function
    await getUserDetails(mockReq, mockRes);


    // Verify the error handling
    expect(axios.get).toHaveBeenCalledWith('https://api.github.com/users/invalid');
    expect(mockRes.status).toHaveBeenCalledWith(500);
    expect(mockRes.json).toHaveBeenCalledWith({ error: 'Error fetching user details' });
    
    
    // Snapshot the error response
    expect(mockRes.json).toMatchSnapshot();
  });
});
