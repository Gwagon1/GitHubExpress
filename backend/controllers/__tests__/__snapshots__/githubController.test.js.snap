// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUserDetails handles errors gracefully 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "error": "Error fetching user details",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
